runtime/src/main/java/org/corfudb/runtime/CorfuRuntime.java

```diff
@@ -366,45 +559,143 @@ private CorfuRuntime(@Nonnull CorfuRuntimeParameters parameters) {
     }
 
     /**
-     * Shuts down the CorfuRuntime.
-     * Stops async tasks from fetching the layout.
-     * Cannot reuse the runtime once shutdown is called.
+     * Shutdown the runtime, stop asynchronous tasks and free resources.
+     *
+     * <p>Once the runtime is shutdown, requests will fail with an exception and the
+     * runtime cannot be reused. {@link this#isShutdown()} will return true to reflect
+     * that the runtime is shutdown.
+     *
+     * <p>This method is synchronized to prevent multiple callers from initiating
+     * a shutdown, and is therefore thread-safe (the runtime will only be shutdown once).
      */
-    public void shutdown() {
+    public synchronized void shutdown() {
+        // If we are already shutdown, throw a shutdown error.
+        if (isShutdown) {
+            return;
+        }
 
-        // Stopping async task from fetching layout.
+        log.info("shutdown: Corfu runtime shutting down");
+        // Set the shutdown flag to true. This stops any active layout acquisition
+        // from retrying.
         isShutdown = true;
-        if (layout != null) {
-            try {
-                layout.cancel(true);
-            } catch (Exception e) {
-                log.error("Runtime shutting down. Exception in terminating fetchLayout: {}", e);
-            }
+
+        // Generate a new layout which completes exceptionally. This stops any thread
+        // which is trying to get a layout.
+        layout = new CompletableFuture<>();
+        layout.completeExceptionally(new RuntimeShutdownError());
+
+        // Stop each active router. This stops network RPC.
+        nodeRouters.forEach((name, router) -> router.stop());
+
+        // Shutdown our executors, if requested.
+        if (parameters.shutdownIoExecutor) {
+            ioExecutor.shutdown();
         }
-        stop(true);
+
+        // Shutdown the event loop, if requested.
+        if (parameters.shutdownNettyEventLoop) {
+            nettyEventLoop.shutdownGracefully().syncUninterruptibly();
+        }
+
+
+        log.info("shutdown: Corfu runtime successfully shutdown");
     }
 
-    /**
-     * Stop all routers associated with this runtime & disconnect them.
+    /** Connect to the Corfu cluster.
+     *
+     * <p>When the method returns, the Corfu cluster is connected and the {@link CorfuRuntime}
+     * can access Corfu cluster services.
+     *
+     * @return  This {@link CorfuRuntime}, to support chaining calls.
      */
-    public void stop() {
-        stop(false);
+    public CorfuRuntime connect() {
+        if (!layout.isDone()) {
+            invalidateLayout();
```

Comment:
The naming convention in this is not intuitive. Since invalidateLayout actually restores a new layout it should be called something better.